File: App/Resources/AlgorithmDetails/basenmaxheapsort/basenmaxheapsort.swift
let base = 4

func siftDown(_ arr: inout [Int], _ node: Int, _ stop: Int) {
    let left = node * base + 1
    if left >= stop {
        return
    }
    var maxIndex = left
    var i = left + 1
    while i < left + base && i < stop {
        if arr[maxIndex] < arr[i] {
            maxIndex = i
        }
        i += 1
    }
    if arr[node] < arr[maxIndex] {
        arr.swapAt(node, maxIndex)
        siftDown(&arr, maxIndex, stop)
    }
}

func sort(_ arr: inout [Int]) {
    let n = arr.count
    var i = n - 1
    while i >= 0 {
        siftDown(&arr, i, n)
        i -= 1
    }
    var end = n - 1
    while end > 0 {
        arr.swapAt(0, end)
        siftDown(&arr, 0, end)
        end -= 1
    }
}

var array: [Int] = [
    0, 39, 21, 62, 91, 77, 14, 23,
    90, 69, 51, 81, 68, 83, 32, 56,
]
sort(&array)
print(array)

File: App/Resources/AlgorithmDetails/basenmaxheapsort/basenmaxheapsort.c
#include <stdio.h>
#include <stdlib.h>

#define BASE 4

int array[16] = {0, 39, 21, 62, 91, 77, 14, 23, 90, 69, 51, 81, 68, 83, 32, 56};

void swap(int *a, int *b) {
  int t = *a;
  *a = *b;
  *b = t;
}

void printList(int items[], int size) {
  for (int i = 0; i < size; i++) {
    if (i == 0) {
      printf("[%d, ", items[i]);
    } else if (i != size - 1) {
      printf("%d, ", items[i]);
    } else {
      printf("%d]", items[i]);
    }
  }
}

void siftDown(int arr[], int node, int stop) {
  int left = node * BASE + 1;
  if (left >= stop) {
    return;
  }
  int maxIndex = left;
  for (int i = left + 1; i < left + BASE && i < stop; i++) {
    if (arr[maxIndex] < arr[i]) {
      maxIndex = i;
    }
  }
  if (arr[node] < arr[maxIndex]) {
    swap(&arr[node], &arr[maxIndex]);
    siftDown(arr, maxIndex, stop);
  }
}

void sort(int arr[], int n) {
  for (int i = n - 1; i >= 0; i--) {
    siftDown(arr, i, n);
  }
  for (int end = n - 1; end > 0; end--) {
    swap(&arr[0], &arr[end]);
    siftDown(arr, 0, end);
  }
}

int main(int argc, char *argv[]) {
  int size = sizeof(array) / sizeof(array[0]);
  sort(array, size);
  printList(array, size);
  return 0;
}
